flink/src/main/java/org/apache/iceberg/flink/sink/compact/CompactFileGenerator.java

```diff
@@ -0,0 +1,243 @@
+        table.properties(),
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES,
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES_DEFAULT);
+    long targetMergeFileSize = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.WRITE_COMPACT_TARGET_FILE_SIZE_BYTES,
+        targetFileSize);
+    this.targetSizeInBytes = Math.min(splitSize, targetMergeFileSize);
+    this.splitLookback = PropertyUtil.propertyAsInt(
+        table.properties(),
+        TableProperties.SPLIT_LOOKBACK,
+        TableProperties.SPLIT_LOOKBACK_DEFAULT);
+    this.splitOpenFileCost = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.SPLIT_OPEN_FILE_COST,
+        TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT);
+  }
+
+  @Override
+  public void processElement(StreamRecord<EndCheckpoint> element) throws Exception {
+    EndCheckpoint endCheckpoint = element.getValue();
+    LOG.info("Received an EndCheckpoint {}, begin to compute CompactionUnit", endCheckpoint.getCheckpointId());
+    emit(new EndCheckpoint(
+        endCheckpoint.getCheckpointId(),
+        getRuntimeContext().getIndexOfThisSubtask(),
+        getRuntimeContext().getIndexOfThisSubtask())
+    );
+
+    if (SmallFileUtil.shouldMergeSmallFiles(table)) {
+      table.refresh();
+      AtomicInteger index = new AtomicInteger();
+      List<CompactionUnit> tasks = getCombinedScanTasks()
+          .stream()
+          .map(combineTask -> new CompactionUnit(combineTask, index.getAndIncrement()))
+          .collect(Collectors.toList());
+
+      if (!tasks.isEmpty()) {
+        tasks.forEach(task -> {
+          emit(task);
+          LOG.info("Emit CompactionUnit(id: {}, files: {}, total: {}, checkpoint: {})",
+              task.getUnitId(),
+              Joiner.on(", ").join(task.getCombinedScanTask().files().stream()
+                  .map(t -> t.file().content() + "=>" + t.file().fileSizeInBytes() / 1024 / 1024 + "M")
+                  .collect(Collectors.toList())),
+              task.getCombinedScanTask().files().size(),
+              endCheckpoint.getCheckpointId());
+        });
+        LOG.info("Summary: (checkpoint: {}, total CompactionUnit: {})", endCheckpoint.getCheckpointId(), tasks.size());
+
+        // broadcast emit checkpoint barrier
+        emit(new EndCompaction(endCheckpoint.getCheckpointId(), startingSnapshotId));
+      } else {
+        LOG.warn("CombinedScanTasks is empty, no files need to be compacted");
+      }
+    }
+  }
+
+  private List<CombinedScanTask> getCombinedScanTasks() {
+    CloseableIterable<FileScanTask> fileScanTasks = null;
+    if (table.currentSnapshot() == null) {
+      return emptyList();
+    }
+
+    long startingSnapshot = table.currentSnapshot().snapshotId();
+    try {
+      fileScanTasks = table.newScan()
+          .useSnapshot(startingSnapshot)
+          .caseSensitive(caseSensitive)
+          .ignoreResiduals()
+          .filter(filter)
+          .planFiles();
```

Comment:
I don't think it is a good idea to attempt to compact the entire table.

The design that we've talked about for this feature is to emit the `DataFile` instances from the committer after a successful commit. Then add a `keyBy` to group the data files by partition in parallel tasks and finally to have each task compact the data files when enough data arrives. Then that compaction writer sends a list of data files to replace and the rewritten data file to another committer that also commits on checkpoint boundaries. Any commit that fails is simply ignored at that point.

I think that's a better model than performing a full table scan.